### PR TITLE
Another batch of fuzzing bugs

### LINKED ISSUE #7358: Fuzzer bug test_syntax "after formatting, the source code no longer parsed"
Run on @joshuawarner32's branch for #7353

```
$ cargo run --bin minimize expr todo
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/minimize expr todo`
Error found: Reparse failed: Expr(Record(End(@0), @0), @0)
Proceeding with minimization
Successfully minimized, new length: 72
Successfully minimized, new length: 71
Successfully minimized, new length: 64
Successfully minimized, new length: 58
Successfully minimized, new length: 53
Successfully minimized, new length: 48
Successfully minimized, new length: 43
Successfully minimized, new length: 39
Successfully minimized, new length: 36
Successfully minimized, new length: 33
Successfully minimized, new length: 31
Successfully minimized, new length: 29
Successfully minimized, new length: 28
Successfully minimized, new length: 27
Successfully minimized, new length: 26
No more minimizations found
Final result:
M{s?s{J&
}}{s?s{J&
}}:p
y
```
```
thread '<unnamed>' panicked at /Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/src/test_helpers.rs:333:13:
After formatting, the source code no longer parsed!

Parse error was: Expr(Record(End(@119), @100), @0)

The original code was:

Mxpsat{pdes?s{J&ta,rvm
?r}}Mxpsat{pdes?s{J&ta,rvm
?r}}Aqelntp:op

OAqelntp:op

O=
Fs
Aqyqy


The code that failed to parse:

Mxpsat { pdes ? s
        { J &
            ta,
            rvm
            ? r,
        } } Mxpsat { pdes ? s
        { J &
            ta,
            rvm
            ? r,
        } } Aqelntp : op

OAqelntp : op

O =
    Fs
Aqyqy

The original ast was:

Expr(
    @0-90 SpaceAfter(
        Defs(
            Defs {
                tags: [
                    EitherIndex(0),
                    EitherIndex(1),
                    EitherIndex(2147483648),
                ],
                regions: [
                    @0-64,
                    @66-77,
                    @79-84,
                ],
                space_before: [
                    Slice<roc_parse::ast::CommentOrNewline> { start: 0, length: 0 },
                    Slice<roc_parse::ast::CommentOrNewline> { start: 0, length: 2 },
                    Slice<roc_parse::ast::CommentOrNewline> { start: 2, length: 2 },
                ],
                space_after: [
                    Slice<roc_parse::ast::CommentOrNewline> { start: 0, length: 0 },
                    Slice<roc_parse::ast::CommentOrNewline> { start: 2, length: 0 },
                    Slice<roc_parse::ast::CommentOrNewline> { start: 4, length: 0 },
                ],
                spaces: [
                    Newline,
                    Newline,
                    Newline,
                    Newline,
                ],
                type_defs: [
                    Alias {
                        header: TypeHeader {
                            name: @0-6 "Mxpsat",
                            vars: [
                                @6-27 RecordDestructure(
                                    [
                                        @7-26 OptionalField(
                                            "pdes",
                                            @12-26 Apply(
                                                @12-13 Var {
                                                    module_name: "",
                                                    ident: "s",
                                                },
                                                [
                                                    @13-26 RecordUpdate {
                                                        update: @14-15 Tag(
                                                            "J",
                                                        ),
                                                        fields: [
                                                            @16-18 LabelOnly(
                                                                @16-18 "ta",
                                                            ),
                                                            @19-25 OptionalValue(
                                                                @19-22 "rvm",
                                                                [
                                                                    Newline,
                                                                ],
                                                                @24-25 Var {
                                                                    module_name: "",
                                                                    ident: "r",
                                                                },
                                                            ),
                                                        ],
                                                    },
                                                ],
                                                Space,
                                            ),
                                        ),
                                    ],
                                ),
                                @27-33 Tag(
                                    "Mxpsat",
                                ),
                                @33-54 RecordDestructure(
                                    [
                                        @34-53 OptionalField(
                                            "pdes",
                                            @39-53 Apply(
                                                @39-40 Var {
                                                    module_name: "",
                                                    ident: "s",
                                                },
                                                [
                                                    @40-53 RecordUpdate {
                                                        update: @41-42 Tag(
                                                            "J",
                                                        ),
                                                        fields: [
                                                            @43-45 LabelOnly(
                                                                @43-45 "ta",
                                                            ),
                                                            @46-52 OptionalValue(
                                                                @46-49 "rvm",
                                                                [
                                                                    Newline,
                                                                ],
                                                                @51-52 Var {
                                                                    module_name: "",
                                                                    ident: "r",
                                                                },
                                                            ),
                                                        ],
                                                    },
                                                ],
                                                Space,
                                            ),
                                        ),
                                    ],
                                ),
                                @54-61 Tag(
                                    "Aqelntp",
                                ),
                            ],
                        },
                        ann: @62-64 BoundVariable(
                            "op",
                        ),
                    },
                    Alias {
                        header: TypeHeader {
                            name: @66-74 "OAqelntp",
                            vars: [],
                        },
                        ann: @75-77 BoundVariable(
                            "op",
                        ),
                    },
                ],
                value_defs: [
                    Body(
                        @79-80 Tag(
                            "O",
                        ),
                        @82-84 SpaceBefore(
                            Tag(
                                "Fs",
                            ),
                            [
                                Newline,
                            ],
                        ),
                    ),
                ],
            },
            @85-90 SpaceBefore(
                Tag(
                    "Aqyqy",
                ),
                [
                    Newline,
                ],
            ),
        ),
        [
            Newline,
        ],
    ),
)


note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
==2254== ERROR: libFuzzer: deadly signal
NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 1 CopyPart-; base unit: 03c90bd5fa52f03b184d0dbd5bd88f21a43ba352
0x4d,0x78,0x70,0x73,0x61,0x74,0x7b,0x70,0x64,0x65,0x73,0x3f,0x73,0x7b,0x4a,0x26,0x74,0x61,0x2c,0x72,0x76,0x6d,0xa,0x3f,0x72,0x7d,0x7d,0x4d,0x78,0x70,0x73,0x61,0x74,0x7b,0x70,0x64,0x65,0x73,0x3f,0x73,0x7b,0x4a,0x26,0x74,0x61,0x2c,0x72,0x76,0x6d,0xa,0x3f,0x72,0x7d,0x7d,0x41,0x71,0x65,0x6c,0x6e,0x74,0x70,0x3a,0x6f,0x70,0xa,0xa,0x4f,0x41,0x71,0x65,0x6c,0x6e,0x74,0x70,0x3a,0x6f,0x70,0xa,0xa,0x4f,0x3d,0xa,0x46,0x73,0xa,0x41,0x71,0x79,0x71,0x79,0xa,
Mxpsat{pdes?s{J&ta,rvm\012?r}}Mxpsat{pdes?s{J&ta,rvm\012?r}}Aqelntp:op\012\012OAqelntp:op\012\012O=\012Fs\012Aqyqy\012
artifact_prefix='/Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/'; Test unit written to /Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-babf96565ee4a16d60f3f6378048c453c673b1f4
Base64: TXhwc2F0e3BkZXM/c3tKJnRhLHJ2bQo/cn19TXhwc2F0e3BkZXM/c3tKJnRhLHJ2bQo/cn19QXFlbG50cDpvcAoKT0FxZWxudHA6b3AKCk89CkZzCkFxeXF5Cg==
stat::number_of_executed_units: 627430
stat::average_exec_per_sec:     20914
stat::new_units_added:          11069
stat::slowest_unit_time_sec:    0
stat::peak_rss_mb:              432
INFO: exiting: 19712 time: 255s
```

### LINKED ISSUE #7364: Fuzzer bug test_syntax "Formatting bug; formatting is not stable."
`commit 2389040b397f42de928f233b2ac2b34a0e8e875b`

[crash-18822522844fd089f91fd3ebff71116a00b67ffe copy.txt](https://github.com/user-attachments/files/18133105/crash-18822522844fd089f91fd3ebff71116a00b67ffe.copy.txt)

```
$ cargo run --bin minimize expr /Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-18822522844fd089f91fd3ebff71116a00b67ffe
    Finished dev [unoptimized + debuginfo] target(s) in 0.39s
     Running `target/debug/minimize expr /Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-18822522844fd089f91fd3ebff71116a00b67ffe`
Error found: Formatting not stable
Proceeding with minimization
Successfully minimized, new length: 66
Successfully minimized, new length: 65
Successfully minimized, new length: 41
thread 'main' panicked at crates/compiler/parse/src/ast.rs:2354:1:
not yet implemented
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Successfully minimized, new length: 33
Successfully minimized, new length: 28
Successfully minimized, new length: 24
Successfully minimized, new length: 20
Successfully minimized, new length: 17
Successfully minimized, new length: 16
No more minimizations found
Final result:
1,C(#
0
)#
<-F
t
```

### LINKED ISSUE #7365: Fuzzer bug test_syntax `Reparse failed: Expr(IndentEnd(@0), @0)`
`commit 2389040b397f42de928f233b2ac2b34a0e8e875b`
```
$ cargo run --bin minimize expr crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-a1eadc84a3e2e67dbfd15c9b4e693c6b183bfb83
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
     Running `target/debug/minimize expr crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-a1eadc84a3e2e67dbfd15c9b4e693c6b183bfb83`
Error found: Reparse failed: Expr(IndentEnd(@0), @0)
Proceeding with minimization
Successfully minimized, new length: 111
Successfully minimized, new length: 110
Successfully minimized, new length: 109
Successfully minimized, new length: 108
Successfully minimized, new length: 107
Successfully minimized, new length: 106
Successfully minimized, new length: 105
Successfully minimized, new length: 104
Successfully minimized, new length: 103
Successfully minimized, new length: 102
Successfully minimized, new length: 101
Successfully minimized, new length: 100
Successfully minimized, new length: 99
Successfully minimized, new length: 98
Successfully minimized, new length: 41
Successfully minimized, new length: 33
Successfully minimized, new length: 27
Successfully minimized, new length: 25
Successfully minimized, new length: 24
No more minimizations found
Final result:
A implements q:U
 e->p
d
```
[crash-a1eadc84a3e2e67dbfd15c9b4e693c6b183bfb83 copy.txt](https://github.com/user-attachments/files/18133147/crash-a1eadc84a3e2e67dbfd15c9b4e693c6b183bfb83.copy.txt)

### LINKED ISSUE #7366: Fuzzer bug test_syntax `Reparse failed: Expr(Ability(DemandAlignment(0, @0), @0), @0)`
`commit 2389040b397f42de928f233b2ac2b34a0e8e875b`

[crash-7008675616a8a2c27807adb97b111dd65f68ef6e copy.txt](https://github.com/user-attachments/files/18133236/crash-7008675616a8a2c27807adb97b111dd65f68ef6e.copy.txt)

```
$ cargo run --bin minimize expr /Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-7008675616a8a2c27807adb97b111dd65f68ef6e
    Finished dev [unoptimized + debuginfo] target(s) in 0.39s
     Running `target/debug/minimize expr /Users/luke/Documents/GitHub/roc/crates/compiler/test_syntax/fuzz/artifacts/fuzz_expr/crash-7008675616a8a2c27807adb97b111dd65f68ef6e`
Error found: Reparse failed: Expr(Ability(DemandAlignment(0, @0), @0), @0)
Proceeding with minimization
Successfully minimized, new length: 30
Successfully minimized, new length: 23
Successfully minimized, new length: 20
Successfully minimized, new length: 18
No more minimizations found
Final result:
{}=O{}implements
a
```

### PR DESCRIPTION
**Pull request by joshuawarner32**
- **Don't remove empty record assignment if body looks like an expect**
- **Ensure we don't over-indent in type headers (fixes #7358)**
- **Generalize multiline string handling in patterns**
- **lift spaces in backpassing (fixes #7364)**
- **Force initial newline in multiline ability def (fixes #7365)**
- **Disallow eliding empty record destructure assignment if expr looks like an implements clause (fixes #7366)**

### CODE REVIEW

File: crates/compiler/test_syntax/src/minimize.rs [hunk 1]
```diff
@@ -81,7 +95,8 @@ fn round_trip_once(input: Input<'_>) -> Option<String> {
 
     let actual = match input.parse_in(&arena) {
         Ok(a) => a,
-        Err(e) => return Some(format!("Initial parse failed: {:?}", e.normalize(&arena))),
+        Err(_e) => return None,
+        //     Err(e) => return Some(format!("Initial parse failed: {:?}", e.normalize(&arena))),
```
Comment (lukewilliamboswell) [inline crates/compiler/test_syntax/src/minimize.rs:99]: Commented out line

File: crates/compiler/test_syntax/tests/test_snapshots.rs [hunk 2]
```diff
@@ -730,6 +765,13 @@ mod test_snapshots {
                 true
             }
 
+            "parens_newlines_before_as" => {
+                // Found unbound type variables {66}
+                //  in type alias `Test.IdentId(0)` [] [] : <66>
+                // Location: crates/compiler/can/src/scope.rs:541:13
```
Comment (lukewilliamboswell) [inline crates/compiler/test_syntax/tests/test_snapshots.rs:771]: What's going on here?

File: examples/virtual-dom-wip/platform/Json.roc [hunk 3]
```diff
@@ -1205,79 +1205,79 @@ expect
...
+                # Decode the json value
+                (
+                    fieldName = objectName
+
+                    # Retrieve value decoder for the current field
+                    when stepField recordState fieldName is
+                        Skip ->
+                            # TODO This doesn't seem right, shouldn't we eat
+                            # the remaining json object value bytes if we are skipping this
+                            # field?
```
Comment (lukewilliamboswell) [inline examples/virtual-dom-wip/platform/Json.roc:1239]: TODO

**Review by lukewilliamboswell — DISMISSED** [review]
Noticed a few things and left comments on those.

File: crates/compiler/test_syntax/tests/test_snapshots.rs (hunk 2 above)
Comment (joshuawarner32) [reply crates/compiler/test_syntax/tests/test_snapshots.rs:771]: This is the panic in canonicalization that these two tests run into. I'll get to this eventually, but getting fuzz clean is higher on my list.

Or, that could be an interesting first issue for an enterprising individual!

File: examples/virtual-dom-wip/platform/Json.roc (hunk 3 above)
Comment (joshuawarner32) [reply examples/virtual-dom-wip/platform/Json.roc:1239]: This was an existing TODO; all of this code just got unindented 4 spaces.

**Review by lukewilliamboswell — APPROVED** [review]

**Pull request merged**